File: packages/sdk-dotnet/src/Rollgate.SDK/TelemetryCollector.cs
using System.Collections.Concurrent;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace Rollgate.SDK;

/// <summary>
/// Collects flag evaluation telemetry and flushes periodically to the server.
/// Thread-safe via ConcurrentDictionary and Interlocked operations.
/// </summary>
public class TelemetryCollector : IDisposable
{
    private readonly string _endpoint;
    private readonly string _apiKey;
    private readonly int _flushIntervalMs;
    private readonly int _maxBufferSize;
    private readonly HttpClient _httpClient;
    private readonly ConcurrentDictionary<string, TelemetryEvalStats> _evaluations = new();
    private CancellationTokenSource? _flushCts;
    private long _periodStartTicks;

    /// <summary>
    /// Tracks evaluation counts for a single flag. Thread-safe via Interlocked.
    /// </summary>
    private class TelemetryEvalStats
    {
        private int _total;
        private int _trueCount;
        private int _falseCount;

        public int Total => _total;
        public int TrueCount => _trueCount;
        public int FalseCount => _falseCount;

        public void Record(bool value)
        {
            Interlocked.Increment(ref _total);
            if (value)
                Interlocked.Increment(ref _trueCount);
            else
                Interlocked.Increment(ref _falseCount);
        }
    }

    public TelemetryCollector(
        string endpoint,
        string apiKey,
        int flushIntervalMs,
        int maxBufferSize,
        HttpClient httpClient)
    {
        _endpoint = endpoint;
        _apiKey = apiKey;
        _flushIntervalMs = flushIntervalMs;
        _maxBufferSize = maxBufferSize;
        _httpClient = httpClient;
        _periodStartTicks = Environment.TickCount64;
    }

    /// <summary>
    /// Start the periodic flush task.
    /// </summary>
    public void Start()
    {
        _flushCts = new CancellationTokenSource();
        var ct = _flushCts.Token;
        _ = Task.Run(async () =>
        {
            while (!ct.IsCancellationRequested)
            {
                try { await Task.Delay(_flushIntervalMs, ct); }
                catch (OperationCanceledException) { return; }
                try { await FlushAsync(ct); }
                catch { /* ignore flush errors */ }
            }
        }, ct);
    }

    /// <summary>
    /// Stop the periodic flush and perform a final flush.
    /// </summary>
    public async Task StopAsync()
    {
        _flushCts?.Cancel();
        try { await FlushAsync(); }
        catch { /* best-effort final flush */ }
    }

    /// <summary>
    /// Record a flag evaluation result.
    /// </summary>
    public void RecordEvaluation(string flagKey, bool result)
    {
        var stats = _evaluations.GetOrAdd(flagKey, _ => new TelemetryEvalStats());
        stats.Record(result);

        // Auto-flush if buffer exceeds max size
        if (_evaluations.Count >= _maxBufferSize)
        {
            _ = Task.Run(async () => { try { await FlushAsync(); } catch { } });
        }
    }

    /// <summary>
    /// Flush all buffered telemetry to the server.
    /// </summary>
    public async Task FlushAsync(CancellationToken ct = default)
    {
        // Snapshot and clear atomically
        var snapshot = new Dictionary<string, TelemetryEvalStats>();
        foreach (var key in _evaluations.Keys)
        {
            if (_evaluations.TryRemove(key, out var stats))
                snapshot[key] = stats;
        }

        if (snapshot.Count == 0) return;

        var now = Environment.TickCount64;
        var periodMs = now - _periodStartTicks;
        _periodStartTicks = now;

        var evaluations = new Dictionary<string, object>();
        foreach (var kv in snapshot)
        {
            evaluations[kv.Key] = new Dictionary<string, int>
            {
                ["total"] = kv.Value.Total,
                ["true"] = kv.Value.TrueCount,
                ["false"] = kv.Value.FalseCount,
            };
        }

        var payload = new Dictionary<string, object>
        {
            ["evaluations"] = evaluations,
            ["period_ms"] = periodMs,
        };

        var json = JsonSerializer.Serialize(payload);
        var content = new StringContent(json, Encoding.UTF8, "application/json");

        var request = new HttpRequestMessage(HttpMethod.Post, _endpoint) { Content = content };
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);

        try
        {
            using var requestCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            requestCts.CancelAfter(TimeSpan.FromSeconds(10));
            var response = await _httpClient.SendAsync(request, requestCts.Token);

            if (!response.IsSuccessStatusCode)
            {
                ReBuffer(snapshot);
            }
        }
        catch
        {
            ReBuffer(snapshot);
            throw;
        }
    }

    /// <summary>
    /// Get current buffer statistics.
    /// </summary>
    public (int flagCount, int evaluationCount) GetBufferStats()
    {
        var flagCount = _evaluations.Count;
        var evaluationCount = 0;
        foreach (var kv in _evaluations)
        {
            evaluationCount += kv.Value.Total;
        }
        return (flagCount, evaluationCount);
    }

    public void Dispose()
    {
        _flushCts?.Cancel();
        // Best-effort final flush
        try { FlushAsync().GetAwaiter().GetResult(); }
        catch { }
    }

    /// <summary>
    /// Re-buffer stats on flush failure to avoid data loss.
    /// </summary>
    private void ReBuffer(Dictionary<string, TelemetryEvalStats> snapshot)
    {
        foreach (var kv in snapshot)
        {
            _evaluations.AddOrUpdate(
                kv.Key,
                _ => kv.Value,
                (_, existing) =>
                {
                    // Merge: record each evaluation from the failed snapshot into the existing stats
                    for (var i = 0; i < kv.Value.TrueCount; i++) existing.Record(true);
                    for (var i = 0; i < kv.Value.FalseCount; i++) existing.Record(false);
                    return existing;
                });
        }
    }
}

File: packages/sdk-java/src/main/java/io/rollgate/TelemetryCollector.java
package io.rollgate;

import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.*;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Collects and batches flag evaluation telemetry data.
 * Periodically flushes aggregated evaluation stats to the server.
 */
public class TelemetryCollector {

    private final String endpoint;
    private final String apiKey;
    private final long flushIntervalMs;
    private final int maxBufferSize;
    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;

    private final ConcurrentHashMap<String, TelemetryEvalStats> evaluations = new ConcurrentHashMap<>();
    private final AtomicBoolean isFlushing = new AtomicBoolean(false);
    private final AtomicInteger totalBuffered = new AtomicInteger(0);

    private ScheduledExecutorService scheduler;
    private long periodStartMs;

    public TelemetryCollector(String endpoint, String apiKey, long flushIntervalMs,
                              int maxBufferSize, OkHttpClient httpClient) {
        this.endpoint = endpoint;
        this.apiKey = apiKey;
        this.flushIntervalMs = flushIntervalMs;
        this.maxBufferSize = maxBufferSize;
        this.httpClient = httpClient;
        this.objectMapper = new ObjectMapper();
        this.periodStartMs = System.currentTimeMillis();
    }

    /**
     * Start the periodic flush scheduler.
     */
    public void start() {
        if (scheduler != null) {
            return;
        }

        scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "rollgate-telemetry");
            t.setDaemon(true);
            return t;
        });

        scheduler.scheduleAtFixedRate(this::flushQuietly, flushIntervalMs, flushIntervalMs, TimeUnit.MILLISECONDS);
    }

    /**
     * Stop the collector and flush remaining telemetry.
     */
    public void stop() {
        if (scheduler != null) {
            scheduler.shutdown();
            try {
                if (!scheduler.awaitTermination(2, TimeUnit.SECONDS)) {
                    scheduler.shutdownNow();
                }
            } catch (InterruptedException e) {
                scheduler.shutdownNow();
                Thread.currentThread().interrupt();
            }
            scheduler = null;
        }
        flushQuietly();
    }

    /**
     * Record a flag evaluation result.
     *
     * @param flagKey The flag key that was evaluated
     * @param result  The evaluation result (true/false)
     */
    public void recordEvaluation(String flagKey, boolean result) {
        if (totalBuffered.get() >= maxBufferSize) {
            // Buffer full, drop oldest by flushing
            flushQuietly();
        }

        evaluations.compute(flagKey, (key, stats) -> {
            if (stats == null) {
                stats = new TelemetryEvalStats();
            }
            stats.record(result);
            return stats;
        });

        totalBuffered.incrementAndGet();
    }

    /**
     * Flush all buffered telemetry to the server.
     */
    public void flush() throws IOException {
        if (!isFlushing.compareAndSet(false, true)) {
            return;
        }

        try {
            if (evaluations.isEmpty()) {
                return;
            }

            // Snapshot and clear
            Map<String, TelemetryEvalStats> snapshot = new HashMap<>();
            for (Map.Entry<String, TelemetryEvalStats> entry : evaluations.entrySet()) {
                snapshot.put(entry.getKey(), entry.getValue().snapshot());
            }
            long periodMs = System.currentTimeMillis() - periodStartMs;

            evaluations.clear();
            totalBuffered.set(0);
            periodStartMs = System.currentTimeMillis();

            // Build payload
            Map<String, Object> evalPayload = new HashMap<>();
            for (Map.Entry<String, TelemetryEvalStats> entry : snapshot.entrySet()) {
                TelemetryEvalStats stats = entry.getValue();
                Map<String, Integer> flagStats = new HashMap<>();
                flagStats.put("total", stats.getTotal());
                flagStats.put("true", stats.getTrueCount());
                flagStats.put("false", stats.getFalseCount());
                evalPayload.put(entry.getKey(), flagStats);
            }

            Map<String, Object> payload = new HashMap<>();
            payload.put("evaluations", evalPayload);
            payload.put("period_ms", periodMs);

            String json = objectMapper.writeValueAsString(payload);

            Request request = new Request.Builder()
                .url(endpoint)
                .addHeader("Authorization", "Bearer " + apiKey)
                .addHeader("Content-Type", "application/json")
                .post(RequestBody.create(json, MediaType.parse("application/json")))
                .build();

            try (Response response = httpClient.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    // Re-buffer on failure
                    for (Map.Entry<String, TelemetryEvalStats> entry : snapshot.entrySet()) {
                        evaluations.merge(entry.getKey(), entry.getValue(), TelemetryEvalStats::merge);
                    }
                    int recount = 0;
                    for (TelemetryEvalStats s : evaluations.values()) {
                        recount += s.getTotal();
                    }
                    totalBuffered.set(recount);
                    throw new IOException("Telemetry flush failed with status " + response.code());
                }
            } catch (IOException e) {
                // Re-buffer on network error (if not already re-buffered)
                if (evaluations.isEmpty()) {
                    for (Map.Entry<String, TelemetryEvalStats> entry : snapshot.entrySet()) {
                        evaluations.merge(entry.getKey(), entry.getValue(), TelemetryEvalStats::merge);
                    }
                    int recount = 0;
                    for (TelemetryEvalStats s : evaluations.values()) {
                        recount += s.getTotal();
                    }
                    totalBuffered.set(recount);
                }
                throw e;
            }
        } finally {
            isFlushing.set(false);
        }
    }

    /**
     * Get buffer statistics.
     *
     * @return int array where [0] = number of distinct flags, [1] = total evaluation count
     */
    public int[] getBufferStats() {
        return new int[]{evaluations.size(), totalBuffered.get()};
    }

    private void flushQuietly() {
        try {
            flush();
        } catch (IOException ignored) {
        }
    }

    /**
     * Aggregated evaluation statistics for a single flag.
     */
    static class TelemetryEvalStats {
        private int total;
        private int trueCount;
        private int falseCount;

        TelemetryEvalStats() {
            this.total = 0;
            this.trueCount = 0;
            this.falseCount = 0;
        }

        TelemetryEvalStats(int total, int trueCount, int falseCount) {
            this.total = total;
            this.trueCount = trueCount;
            this.falseCount = falseCount;
        }

        synchronized void record(boolean result) {
            total++;
            if (result) {
                trueCount++;
            } else {
                falseCount++;
            }
        }

        synchronized int getTotal() {
            return total;
        }

        synchronized int getTrueCount() {
            return trueCount;
        }

        synchronized int getFalseCount() {
            return falseCount;
        }

        synchronized TelemetryEvalStats snapshot() {
            return new TelemetryEvalStats(total, trueCount, falseCount);
        }

        static TelemetryEvalStats merge(TelemetryEvalStats a, TelemetryEvalStats b) {
            return new TelemetryEvalStats(
                a.getTotal() + b.getTotal(),
                a.getTrueCount() + b.getTrueCount(),
                a.getFalseCount() + b.getFalseCount()
            );
        }
    }
}
